cmd/storage-rest-client.go

```diff
@@ -197,6 +199,42 @@ func (client *storageRESTClient) SetDiskID(id string) {
 	client.diskID = id
 }
 
+func (client *storageRESTClient) Latency() (ret int64) {
+	now := time.Now()
+	val, timestamp := client.latency.load()
+	if now.Sub(timestamp) < time.Minute {
+		return val
+	}
+
+	errCh := make(chan error)
+	go func() {
+		_, err := client.ReadAll(context.TODO(), minioMetaTmpBucket, formatConfigFile)
+		errCh <- err
+		close(errCh)
+	}()
+
+	var (
+		latency    int64
+		updateTime time.Time
+	)
+
+	select {
+	case err := <-errCh:
+		updateTime = time.Now()
+		if err == nil {
+			latency = int64(updateTime.Sub(now))
+		} else {
+			latency = -1
+		}
+	case <-time.NewTimer(5 * time.Second).C:
+		updateTime = time.Now()
+		latency = -1
+	}
+
+	client.latency.store(latency, updateTime)
+	return latency
+}
+
```

Comment:
I think we shouldn't measure latency like this, in-fact we should have a mean latency per Disk either remote or local calculated for each operation. 

Such that we know that a drive is slower / network is slower for long periods of time and they can be ignored sufficiently. 

This allows tracking latency for all operations otherwise just reading/hitting `format.json` would actually cause disk head movements which would cause random I/O on the drives.

Reply:
This latency measurement gives an overview of the Disk on how performant it is across actual calls, this way you don't have to measure this out of band i.e as a separate API call. 

This allows for real-time measurement of disk speeds all the time and larger delays / significant drops we could potentially take that drive offline for a bit until the latency becomes acceptable.

Reply:
So basically, the only purpose of Latency() is for getLoadBalancedDisks() to return faster local/remote disks first.

Latency() is only called each x minutes. I am not sure if getting mean values can be useful here.

We can discuss this live anyway.

Reply:
But we cannot really spin the disk heads like this. It can create ridiculous amount of random i/o 

Hot path calls like ListObjects cannot have this kind of dependency on the disk, because it the disk heads move it may slow down listing further.

If needed we should capture two things

- LatencyMean()
- LatencyCurrent()

You may use current latency by capturing the latency of the last call, that happened on the drive - if the last call slower than most then - ignore such drives.

Rely on intelligence obtained by other calls instead of performing extra calls to disk.

Reply:
@harshavardhana it is hard to make a disk to know that it is faster than other disks. One idea I had is to measure disk speed for write/reading. However, for reading/writing objects, this will depend on S3 clients as well.

Another solution is to set disks preference at erasure layer. It means when erasure sets requests to different disks, it will remember the fastest disks that respond first. Let's discuss this online whenever you have time.

Reply:
Read ops will also likely get cached on the receiver side, so you would only really measure network latency+cpu load.

Reply:
> Read ops will also likely get cached on the receiver side,

Yes, correct. I am just not able to find any reliable way to compare disks (local or remote) performance at xl-storage layer.

The only way is to do that at higher level, in erasure level, probably while reading objects.